functorch/csrc/BatchRulesNorm.cpp

```diff
@@ -5,107 +5,292 @@
+  VmapDimVector repeat_sizes(input.dim(), 1);
+  repeat_sizes[1] = batch_size;
+  return input.repeat(repeat_sizes);
+}
+
+template<typename F, F Func>
+std::tuple<Tensor,optional<int64_t>,Tensor,optional<int64_t>,Tensor,optional<int64_t>>
 batch_norm_batch_rule(
     const Tensor& input, optional<int64_t> input_bdim,
-    const c10::optional<Tensor>& weight, optional<int64_t> weight_bdim,
-    const c10::optional<Tensor>& bias, optional<int64_t> bias_bdim,
-    const c10::optional<Tensor>& running_mean, optional<int64_t> running_mean_bdim,
-    const c10::optional<Tensor>& running_var, optional<int64_t> running_var_bdim,
-    bool training, double momentum, double eps, bool cudnn_enabled) {
-  TORCH_INTERNAL_ASSERT(input_bdim.has_value());
-  auto batch_size = input.size(*input_bdim);
-  auto input_ = reshape_dim_into(*input_bdim, /*channels dim*/1, input);
-
-  auto weight_ = maybe_flatten(weight, weight_bdim);
-  auto bias_ = maybe_flatten(bias, bias_bdim);
-  auto running_mean_ = maybe_flatten(running_mean, running_mean_bdim);
-  auto running_var_ = maybe_flatten(running_var, running_var_bdim);
-
-  auto result = at::batch_norm(
-      input_, weight_, bias_, running_mean_, running_var_,
-      training, momentum, eps, cudnn_enabled);
-  return std::make_tuple(reshape_dim_outof(1, batch_size, result), 1);
+    const c10::optional<Tensor>& weight_opt, optional<int64_t> weight_bdim,
+    const c10::optional<Tensor>& bias_opt, optional<int64_t> bias_bdim,
+    const c10::optional<Tensor>& running_mean_opt, optional<int64_t> running_mean_bdim,
+    const c10::optional<Tensor>& running_var_opt, optional<int64_t> running_var_bdim,
+    bool training, double momentum, double eps) {
+  c10::MaybeOwned<Tensor> weight_maybe_owned = at::borrow_from_optional_tensor(weight_opt);
+  const Tensor& weight = *weight_maybe_owned;
+  c10::MaybeOwned<Tensor> bias_maybe_owned = at::borrow_from_optional_tensor(bias_opt);
+  const Tensor& bias = *bias_maybe_owned;
+  c10::MaybeOwned<Tensor> running_mean_maybe_owned = at::borrow_from_optional_tensor(running_mean_opt);
+  auto running_mean_ = *running_mean_maybe_owned;
+  c10::MaybeOwned<Tensor> running_var_maybe_owned = at::borrow_from_optional_tensor(running_var_opt);
+  auto running_var_ = *running_var_maybe_owned;
+
+  if (input_bdim && ((running_mean_.defined() && !running_mean_bdim) || (running_var_.defined() && !running_var_bdim))) {
+    throw std::runtime_error("Batch norm got a batched tensor as input while the running_mean or running_var, which will be updated in place, were not batched.");
+  }
+  auto input_ = input;
+  c10::optional<int64_t> bdim_size;
+  if (input_bdim) {
+    bdim_size = input.size(*input_bdim);
+    input_ = reshape_dim_into(*input_bdim, /*channels dim*/1, input);
+  } else if (running_mean_bdim) {
+    bdim_size = running_mean_.size(*running_mean_bdim);
+    input_ = repeatForUpdatedValues(input_, *bdim_size);
```

Comment:
Yes this is the output shape (repeat will basically give a tensor with a shape that multiplies each dimension by the shape given, like `[input.size(i) * repeat_size[i] for i in range(len(input.shape))]`).

I saw after writing this while I was writing the backwards rule your `ensure_has_bdim` which we can use and then reshape instead but we might not gain a lot since cudnn and miopen still need the input to be contiguous so we might have to make the inputs contiguous again...I'll look at this